File: src/main/java/net/hectus/neobb/util/Utilities.java
package net.hectus.neobb.util;

import com.marcpg.libpg.lang.Translation;
import net.hectus.neobb.game.util.GameManager;
import net.hectus.neobb.player.NeoPlayer;
import org.bukkit.Location;
import org.bukkit.Sound;
import org.bukkit.World;
import org.bukkit.entity.EntityType;
import org.bukkit.entity.Player;
import org.bukkit.event.Cancellable;
import org.jetbrains.annotations.NotNull;

import java.time.*;
import java.util.List;
import java.util.Locale;
import java.util.Random;
import java.util.function.Consumer;
import java.util.function.Predicate;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import static org.bukkit.entity.EntityType.*;

public final class Utilities {
    public static final List<EntityType> ENTITY_TYPES = List.of(
            WITHER_SKELETON, STRAY, HUSK, ZOMBIE_VILLAGER, SKELETON_HORSE, ZOMBIE_HORSE, DONKEY, MULE, EVOKER, VEX,
            VINDICATOR, CREEPER, SKELETON, SPIDER, ZOMBIE, SLIME, GHAST, ZOMBIFIED_PIGLIN, ENDERMAN, CAVE_SPIDER,
            SILVERFISH, BLAZE, MAGMA_CUBE, BAT, WITCH, ENDERMITE, GUARDIAN, SHULKER, PIG, SHEEP, COW, CHICKEN, SQUID,
            WOLF, MOOSHROOM, SNOW_GOLEM, OCELOT, IRON_GOLEM, HORSE, RABBIT, POLAR_BEAR, LLAMA, PARROT, VILLAGER, TURTLE,
            PHANTOM, DROWNED, CAT, PANDA, PILLAGER, RAVAGER, TRADER_LLAMA, WANDERING_TRADER, FOX, BEE, HOGLIN, PIGLIN,
            STRIDER, ZOGLIN, AXOLOTL, GLOW_SQUID, GOAT, ALLAY, FROG, CAMEL, SNIFFER
    );

    public static final Random RANDOM = new Random();
    public static final Pattern NUMBER_PATTERN = Pattern.compile("\\d+");

    public static @NotNull String camelToSnake(@NotNull String input) {
        return input.replaceAll("([a-z])([A-Z]+)", "$1_$2").toLowerCase();
    }

    public static @NotNull String counterFilterName(@NotNull String className) {
        return className.replaceAll("^.?T(?!.*(Function|Usage|Clazz)$)|Function$|Usage$|Clazz$", "");
    }

    public static @NotNull String turnName(@NotNull String className, Locale locale) {
        return Translation.string(locale, "turns." + camelToSnake(counterFilterName(className)));
    }

    public static int extractNumber(String input) {
        Matcher matcher = NUMBER_PATTERN.matcher(input);
        if (matcher.find())
            return Integer.parseInt(matcher.group());
        return 0;
    }

    public static @NotNull Location listToLocation(World world, @NotNull List<Integer> list) {
        return new Location(world, list.get(0), list.get(1), list.get(2));
    }

    public static <T> void loop(T @NotNull [][][] array, boolean allowNull, Consumer<T> action) {
        for (T[][] vX : array) {
            for (T[] vY : vX) {
                for (T v : vY) {
                    if (allowNull || v != null) action.accept(v);
                }
            }
        }
    }

    public static void cancelEvent(@NotNull Cancellable event, Player eventPlayer, boolean started, Predicate<NeoPlayer> additionalPredicate) {
        playerEventAction(eventPlayer, started, additionalPredicate, p -> {
            event.setCancelled(true);
            p.player.playSound(p.player, Sound.ENTITY_VILLAGER_NO, 0.5f, 1.0f);
        });
    }

    public static void playerEventAction(Player eventPlayer, boolean started, Predicate<NeoPlayer> additionalPredicate, Consumer<NeoPlayer> action) {
        NeoPlayer player = GameManager.player(eventPlayer, true);
        if (player != null && (!started || player.game.started()) && additionalPredicate.test(player))
            action.accept(player);
    }

    private static final ZoneId EARLIEST_TIMEZONE = ZoneId.of("Pacific/Honolulu");  // UTC−10
    private static final ZoneId LATEST_TIMEZONE = ZoneId.of("Pacific/Auckland"); // UTC+12

    /**
     * Checks whether it's a weekday (monday-friday) in any major time zone.
     * This is inclusive, meaning it only requires one to be true, to not cause any confusion amongst
     * international players. The time zones that are checked are from UTC-10 to UTC+12, because no major
     * population lives beyond that point. No, Kiritimati is not a major population center.
     * @return {@code true} if it is monday-friday, {@code false} if it is saturday or sunday.
     */
    public static boolean isWeekday() {
        return isWeekday(ZonedDateTime.now(EARLIEST_TIMEZONE).getDayOfWeek()) &&
                isWeekday(ZonedDateTime.now(LATEST_TIMEZONE).getDayOfWeek());
    }

    public static boolean isWeekday(DayOfWeek day) {
        return day != DayOfWeek.SATURDAY && day != DayOfWeek.SUNDAY;
    }

    /**
     * Checks if it is daytime (8am to 8pm) in the UTC time zone.
     * @return {@code true} if it is daytime, {@code false} if it is nighttime.
     */
    public static boolean isDaytime() {
        LocalTime time = ZonedDateTime.now(ZoneOffset.UTC).toLocalTime();
        return time.isAfter(LocalTime.of(8, 0)) && time.isBefore(LocalTime.of(20, 0));
    }
}

File: src/main/kotlin/net/hectus/neobb/util/Utilities.kt
package net.hectus.neobb.util

import com.destroystokyo.paper.ParticleBuilder
import com.marcpg.libpg.display.MinecraftReceiver
import com.marcpg.libpg.display.playSound
import com.marcpg.libpg.storing.Cord
import com.marcpg.libpg.util.toCord
import net.hectus.neobb.game.GameManager
import net.hectus.neobb.player.NeoPlayer
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.format.TextColor
import net.kyori.adventure.text.format.TextDecoration
import org.bukkit.Location
import org.bukkit.Sound
import org.bukkit.entity.EntityType
import org.bukkit.entity.EntityType.*
import org.bukkit.entity.LivingEntity
import org.bukkit.entity.Player
import org.bukkit.event.Cancellable
import org.bukkit.util.Vector
import java.time.*

object Utilities {
    val REMOVED_ENTITY_TYPES = listOf(ENDER_DRAGON, ELDER_GUARDIAN, GIANT, ILLUSIONER, RAVAGER, WITHER)

    val ENTITY_TYPES: List<EntityType> = EntityType.entries.filter { it.entityClass != null && it !in REMOVED_ENTITY_TYPES && LivingEntity::class.java.isAssignableFrom(it.entityClass!!) }

    private val EARLIEST_TIMEZONE: ZoneId = ZoneId.of("Pacific/Honolulu") // UTC−10
    private val LATEST_TIMEZONE: ZoneId = ZoneId.of("Pacific/Auckland") // UTC+12

    /**
     * Checks whether it's a weekday (Monday-Friday) in any major time zone.
     * This is inclusive, meaning it only requires one to be true to not cause any confusion amongst
     * international players. The time zones that are checked are from UTC-10 to UTC+12, because no major
     * population lives beyond that point. No, Kiritimati is not a major population center.
     * @return `true` if it is monday-friday, `false` if it is saturday or sunday.
     */
    fun isWeekday(): Boolean {
        return isWeekday(ZonedDateTime.now(EARLIEST_TIMEZONE).dayOfWeek) &&
                isWeekday(ZonedDateTime.now(LATEST_TIMEZONE).dayOfWeek)
    }

    fun isWeekday(day: DayOfWeek): Boolean = day != DayOfWeek.SATURDAY && day != DayOfWeek.SUNDAY

    /**
     * Checks if it is daytime (8am to 8pm) in the UTC time zone.
     * @return `true` if it is daytime, `false` if it is nighttime.
     */
    fun isDaytime(): Boolean {
        val time = ZonedDateTime.now(ZoneOffset.UTC).toLocalTime()
        return time.isAfter(LocalTime.of(8, 0)) && time.isBefore(LocalTime.of(20, 0))
    }

    fun generateBeam(start: Location, direction: Vector, length: Int, spacing: Double, particle: ParticleBuilder, forPoint: (Location) -> Unit = {}) {
        val dir = direction.clone().normalize().multiply(spacing)
        val current = start.clone()

        for (i in 0 until length) {
            current.add(dir)
            particle.location(current).spawn()
            forPoint(current)
        }
    }

    fun colorTransition(color1: TextColor, color2: TextColor, progress: Number): TextColor = TextColor.lerp(progress.toFloat(), color1, color2)
}

data class Bounds(
    val dimensions: Cord,
    val low: Cord,
    val high: Cord = low + dimensions,
    val center3D: Cord = low + (dimensions / Cord(2.0, 2.0, 2.0)),
    val center2D: Cord = Cord(center3D.x, low.y, center3D.z),
) {
    operator fun contains(cord: Cord): Boolean = cord.inBounds(low, high)
    operator fun contains(loc: Location): Boolean = contains(loc.toCord())
}

fun cancelEvent(event: Cancellable, eventPlayer: Player, requireStarted: Boolean, additionalPredicate: (NeoPlayer) -> Boolean) {
    playerEventAction(eventPlayer, requireStarted, additionalPredicate) { p ->
        event.isCancelled = true
        p.playSound(Sound.ENTITY_VILLAGER_NO, 0.5f)
    }
}

fun playerEventAction(eventPlayer: Player, requireStarted: Boolean, additionalPredicate: (NeoPlayer) -> Boolean = { true }, action: (NeoPlayer) -> Unit) {
    val player = GameManager.player(eventPlayer)
    if (player != null && (!requireStarted || player.game.started) && additionalPredicate.invoke(player))
        action.invoke(player)
}

fun MinecraftReceiver.eachNeoPlayer(action: (NeoPlayer) -> Unit) = eachReceiver {
    if (it is NeoPlayer)
        action(it)
}

fun Component.noItalic(): Component = decoration(TextDecoration.ITALIC, TextDecoration.State.FALSE)
